core/scripts/chaincli/handler/keeper_upkeep_history.go

```diff
@@ -0,0 +1,164 @@
+package handler
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"log"
+	"math/big"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/smartcontractkit/chainlink/core/services/keeper"
+	"github.com/smartcontractkit/chainlink/core/services/keystore/keys/ethkey"
+)
+
+// UpkeepHistory prints the checkUpkeep status and keeper responsibility for a given upkeep in a set block range
+func (k *Keeper) UpkeepHistory(ctx context.Context, upkeepId int64, from, to uint64) {
+	registryAddr, registryClient := k.GetRegistry(ctx)
+
+	// Get positioning constant of the current registry
+	positioningConstant, err := keeper.CalcPositioningConstant(upkeepId, ethkey.EIP55AddressFromAddress(registryAddr))
+	if err != nil {
+		log.Fatal("failed to get positioning constant: ", err)
+	}
+	log.Println("Calculated Positioning Constant for the registry: ", positioningConstant)
+
+	callOpts := bind.CallOpts{Context: ctx}
+
+	log.Println("Fetching registry config")
+	registryConfig, err := registryClient.GetConfig(&callOpts)
+	if err != nil {
+		log.Fatal("failed to fetch registry config: ", err)
+	}
+	blockCountPerTurn := registryConfig.BlockCountPerTurn.Uint64()
+
+	log.Println("Fetching registry keepers list")
+	keepersList, err := registryClient.GetKeeperList(&callOpts)
+	if err != nil {
+		log.Fatal("failed to fetch keepers list: ", err)
+	}
+
+	log.Println("Preparing a batch call request")
+	var reqs []rpc.BatchElem
+	var results []*string
+	var keeperPerBlock []uint64
+	for block := from; block <= to; block++ {
+		keeperIndex := (uint64(positioningConstant) + ((block - (block % blockCountPerTurn)) / blockCountPerTurn)) % uint64(len(keepersList))
+		payload, err := keeper.RegistryABI.Pack("checkUpkeep", big.NewInt(upkeepId), keepersList[keeperIndex])
+		if err != nil {
+			log.Fatal("failed to pack checkUpkeep: ", err)
+		}
+
+		var result string
+		reqs = append(reqs, rpc.BatchElem{
+			Method: "eth_call",
+			Args: []interface{}{
+				map[string]interface{}{
+					"to":   k.cfg.RegistryAddress,
+					"data": hexutil.Bytes(payload),
+				},
+				// The block at which we want to inspect the upkeep state
+				hexutil.EncodeUint64(block),
+			},
+			Result: &result,
+		})
+
+		results = append(results, &result)
+		keeperPerBlock = append(keeperPerBlock, keeperIndex)
+	}
+
+	log.Println("Doing batch call to check upkeeps")
+	if err := k.rpcClient.BatchCallContext(ctx, reqs); err != nil {
```

Comment:
How many concurrent requests would this be able to handle? Should we add some batching here?

Reply:
There should be some restriction of max blocks number. Added.